### PR TITLE
Make KubernetesDir a true constant

### LINKED ISSUE #1531: Make KubernetesDir a true constant
The `KubernetesDir` constant is meant to be a constant under normal kubeadm operational circumstances, however it's a variable.

https://github.com/kubernetes/kubernetes/blob/5a790bce3b222b7fd1ef1225e3b20700c577088a/cmd/kubeadm/app/constants/constants.go#L36

The reason for this is, that it's used in many functions deep down the call stack of kubeadm and this makes code hard to test. Most notably the normal location, to which KubernetesDir is pointing, is fixed and usually only writable by root (`/etc/kubernetes`).

To allow tests to be executed by regular users and to stop interference between different test cases, some test cases override the location of `KubernetesDir` to a private (test case only) location. Hence it's no longer a constant, but a variable.

This poses a couple of potential problems:
1. Tests that forget to override it require root privileges to execute and can trash `/etc/kubernetes`.
2. Tests that override it to a private directory, cannot execute in parallel as mixing of the temp directories that different cases use can cause flakyness.

Hence, the need to turn `KubernetesDir` in a true constant and solve the above problems is obvious. There are a couple of possible approaches ATM:

1. More kubeadm funcs should take `KubernetesDir` as a parameter and not use it directly from the constants package. That way it will be easy for unit tests to override the `KubernetesDir` without doing central overriding.
2. Try to execute certain test cases in a chroot. That way `KubernetesDir` can be used as is, but this could be quite tricky to make it work.

/priority backlog
/kind cleanup
/help

### PR DESCRIPTION
**Pull request by SataQiu**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
Make KubernetesDir a true constant

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#1531

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
thanks @SataQiu 
i like that we no longer have to overwrite KubernetesDir, but i wonder if we can simplify this and not require to pass the KubernetesDir in that many places.

/approve
/priority backlog
@kubernetes/sig-cluster-lifecycle-pr-reviews

**Comment by SataQiu** [comment]
@neolit123 
You are right. Can we set up two functions in `constants.go`? 
For example: `CreateDefaultTempDirForKubeadm()` and `CreateTempDirForKubeadm(kubernetesDir string)`.
Choose whether to pass the `kubernetesDir` parameter as needed.
I think I need some suggestions for improvement.

**Comment by neolit123** [comment]
let's see what the others think. i'm mostly ok with the change.

**Comment by timothysc** [comment]
>More kubeadm funcs should take KubernetesDir as a parameter and not use it directly from the constants package. That way it will be easy for unit tests to override the KubernetesDir without doing central overriding.

IMO there is little gained in this ^, and if anything it complicates the signatures in an unnecessary way.

>Try to execute certain test cases in a chroot. That way KubernetesDir can be used as is, but this could be quite tricky to make it work.

Why wouldn't we do this in the config?  

/assign @rosti

**Comment by rosti** [comment]
Indeed, we can add a synthetic field to the internal version of `ClusterConfiguration` that is not present in the versioned variants of the configs (hence invisible for end users).
Then, during conversion from versioned to internal config, we can pin that to `KubernetesDir`.
That should make users of internal `ClusterConfiguration` less likely to have to pass `KubernetesDir` as a parameter.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 1]
```diff
@@ -104,7 +104,7 @@ func livenessProbe(host string, port int, scheme v1.URIScheme) *v1.Probe {
 }
 
 // CreateStaticPodFiles creates all the requested static pod files.
-func CreateStaticPodFiles(manifestDir string, cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.APIEndpoint, componentNames ...string) error {
+func CreateStaticPodFiles(kubernetesDir, manifestDir string, cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.APIEndpoint, componentNames ...string) error {
```
Comment (rosti) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: It's odd for this to take both `manifestDir` and `kubernetesDir`. Can we avoid that?

File: cmd/kubeadm/app/phases/upgrade/staticpods_test.go [hunk 2]
```diff
@@ -433,8 +433,8 @@ func TestStaticPodControlPlane(t *testing.T) {
 			if err != nil {
 				t.Fatalf("couldn't run NewFakeStaticPodPathManager: %v", err)
 			}
-			defer os.RemoveAll(pathMgr.(*fakeStaticPodPathManager).KubernetesDir())
-			constants.KubernetesDir = pathMgr.(*fakeStaticPodPathManager).KubernetesDir()
+			tmpKubernetesDir := pathMgr.(*fakeStaticPodPathManager).KubernetesDir()
```
Comment (rosti) [inline cmd/kubeadm/app/phases/upgrade/staticpods_test.go]: `StaticPodPathManager` should have a `KubernetesDir()` method. We should then be able to initialize a path manager with a different KubernetesDir for tests.

**Review by rosti — COMMENTED** [review]
Thanks for tackling this @SataQiu !
That's certainly a step in the right direction. 
Indeed we should try to reduce the impact of `kubernetesDir` parameters by re-purposing other params such as `manifestsDir`, path manager and/or `ClusterConfiguration`.
Let's investigate those paths too and see where they can lead us.

File: cmd/kubeadm/app/phases/upgrade/staticpods_test.go (hunk 2 above)
Comment (SataQiu) [reply cmd/kubeadm/app/phases/upgrade/staticpods_test.go]: @rosti Sorry for the delay. Thanks for your suggestion. I'll try it in this direction.

**Comment by SataQiu** [comment]
@neolit123 @rosti Thanks for your feedback.
I have done some cleanup work for `KubernetesDir`.
But I am still not sure if we need to change the `ClusterConfiguration`. And how to change it?

File: cmd/kubeadm/app/constants/constants.go [hunk 3]
```diff
@@ -33,11 +33,11 @@ import (
 	"k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
 )
 
-// KubernetesDir is the directory Kubernetes owns for storing various configuration files
-// This semi-constant MUST NOT be modified during runtime. It's a variable solely for use in unit testing.
-var KubernetesDir = "/etc/kubernetes"
-
 const (
+	// KubernetesDir is the directory Kubernetes owns for storing various configuration files
+	// This semi-constant MUST NOT be modified during runtime. It's a variable solely for use in unit testing.
```
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go]: we should remove the last two sentences:
`// This semi-constant MUST NOT be modified during runtime. It's a variable solely for use in unit testing.`

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 4]
```diff
@@ -111,9 +111,11 @@ func CreateStaticPodFiles(manifestDir string, cfg *kubeadmapi.ClusterConfigurati
 		return err
 	}
 
+	kubernetesDir := filepath.Dir(manifestDir)
```
Comment (rosti) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Ok, this is not what I expect here. What we should do is to switch the `manifestDir` parameter of `CreateStaticPodFiles` to `kubernetesDir` and then use `GetStaticPodDirectory` to obtain the dir.
Unfortunately, that may require changes in the path manager interface and this is out of scope of this PR.
Hence, what we need to do here is to pass `kubeadmconstants.KubernetesDir` to `GetStaticPodSpecs` below, add a TODO for flipping `CreateStaticPodFiles` to taking `kubernetesDir` and open an issue with `help wanted` label.

File: cmd/kubeadm/app/phases/upgrade/staticpods.go [hunk 5]
```diff
@@ -90,15 +91,16 @@ func NewKubeStaticPodPathManager(realDir, tempDir, backupDir, backupEtcdDir stri
 
 // NewKubeStaticPodPathManagerUsingTempDirs creates a new instance of KubeStaticPodPathManager with temporary directories backing it
 func NewKubeStaticPodPathManagerUsingTempDirs(realManifestDir string, saveManifestsDir, saveEtcdDir bool) (StaticPodPathManager, error) {
-	upgradedManifestsDir, err := constants.CreateTempDirForKubeadm("kubeadm-upgraded-manifests")
+	kubernetesDir := filepath.Dir(realManifestDir)
```
Comment (rosti) [inline cmd/kubeadm/app/phases/upgrade/staticpods.go]: Again, I don't like this one, you need a top-down approach for this one.
Please, sync up with @ereslibre as you are intersecting with one of his PRs on this part of the code.

**Review by rosti — COMMENTED** [review]
Thanks for the iteration on this one @SataQiu !
This one looks way better than the first one. I like that this one touches only portions of the code, that are needed to make `KubernetesDir` a constant ATM.
There are, however, a couple of instances where a bottom-up approach is used, rather than top-down one.
I am not sure, that you need the synthetic field inside the internal `ClusterConfiguration` for now, so let's leave that for a moment.
Please, sync with @ereslibre as he has made a PR, that is intersecting with this one.

File: cmd/kubeadm/app/phases/upgrade/staticpods.go (hunk 5 above)
Comment (ereslibre) [reply cmd/kubeadm/app/phases/upgrade/staticpods.go]: For completion, the PR that @rosti is referring to is: https://github.com/kubernetes/kubernetes/pull/76929

**Comment by SataQiu** [comment]
@rosti @neolit123 
I chose a new direction to solve this problem.
PTAL!

**Comment by SataQiu** [comment]
@neolit123 Do we have a new way to solve this problem now?  I'm really sorry that I can't attend the online meeting due to the time difference.

File: cmd/kubeadm/app/phases/upgrade/staticpods.go [hunk 6]
```diff
@@ -89,28 +94,34 @@ func NewKubeStaticPodPathManager(realDir, tempDir, backupDir, backupEtcdDir stri
...
 }
 
 // MoveFile should move a file from oldPath to newPath
 func (spm *KubeStaticPodPathManager) MoveFile(oldPath, newPath string) error {
 	return os.Rename(oldPath, newPath)
 }
 
+// KubernetesDir should point to the directory Kubernetes owns for storing various configuration files
+func (spm *KubeStaticPodPathManager) KubernetesDir() string {
+	return spm.kubernetesDir
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/upgrade/staticpods.go]: this is much better.

**Review by dixudx — APPROVED** [review]
/lgtm

**Pull request merged**